File: src/main/java/com/qsoftware/bubbles/gui/QContainer.java
package com.qsoftware.bubbles.gui;

import com.qsoftware.bubbles.common.GraphicsProcessor;

import java.util.ArrayList;

public abstract class QContainer extends QComponent {
    ArrayList<QComponent> components;

    public abstract void renderComponents(GraphicsProcessor ngg);

    public QComponent add(QComponent component) {
        this.components.add(component);
        return component;
    }

    public QComponent remove(QComponent component) {
        this.components.remove(component);
        return component;
    }

    public QComponent removeComponent(int index) {
        return this.components.remove(index);
    }

    public QComponent getComponent(int index) {
        return components.get(index);
    }

    public ArrayList<QComponent> getComponents() {
        return components;
    }
}

File: src/main/kotlin/com/qtech/bubbles/gui/QContainer.kt
package com.qtech.bubbles.gui

import com.qtech.bubbles.graphics.GraphicsProcessor

abstract class QContainer : QComponent() {
    var components: ArrayList<QComponent>? = null
    abstract fun renderComponents(ngg: GraphicsProcessor)
    open fun add(component: QComponent): QComponent {
        components!!.add(component)
        return component
    }

    open fun remove(component: QComponent): QComponent {
        components!!.remove(component)
        return component
    }

    fun removeComponent(index: Int): QComponent {
        return components!!.removeAt(index)
    }

    fun getComponent(index: Int): QComponent {
        return components!![index]
    }
}